packages/noya-state/src/reducers/layerReducer.ts

```diff
@@ -2,28 +2,129 @@ import Sketch from '@sketch-hq/sketch-file-format-ts';
 import produce from 'immer';
 import { AffineTransform, transformRect } from 'noya-geometry';
 import { uuid } from 'noya-renderer';
+import { IndexPath } from 'tree-visit';
 import * as Layers from '../layers';
 import * as Models from '../models';
 import {
-  getBoundingRect,
+  deleteLayers,
   getCurrentPage,
+  getBoundingRect,
   getCurrentPageIndex,
-  getLayerIndexPathsExcludingDescendants,
+  getSymbolsPageIndex,
   getLayerTransformAtIndexPath,
+  getLayerIndexPathsExcludingDescendants,
 } from '../selectors/selectors';
 import { SelectionType, updateSelection } from '../utils/selection';
 import { ApplicationState } from './applicationReducer';
+import { createPage } from './pageReducer';
 
 export type LayerAction =
   | [type: 'deleteLayer', layerId: string | string[]]
   | [type: 'groupLayer', layerId: string | string[], name: string]
   | [type: 'ungroupLayer', layerId: string | string[]]
+  | [type: 'addSymbol', layerId: string | string[], name: string]
   | [
       type: 'selectLayer',
       layerId: string | string[] | undefined,
       selectionType?: SelectionType,
     ];
 
+const getParentLayer = (pages: Sketch.AnyLayer, lastIndexPath: IndexPath) => {
+  return Layers.access(pages, lastIndexPath.slice(0, -1)) as Sketch.Page;
+};
+const addToParentLayer = (
+  pages: Sketch.AnyLayer,
+  selectedIndexPaths: IndexPath[],
+  element: Exclude<Sketch.AnyLayer, { _class: 'page' }>,
+) => {
+  const lastIndexPath = selectedIndexPaths[selectedIndexPaths.length - 1];
+  const parent = getParentLayer(pages, lastIndexPath);
+
+  parent.layers.splice(lastIndexPath[lastIndexPath.length - 1], 0, element);
+};
+
+const getReversedDirectChildrensLayersIndexPaths = (
+  state: ApplicationState,
+  ids: string[],
+) => {
+  return (
+    getLayerIndexPathsExcludingDescendants(state, ids)
+      .filter((indexPath) =>
+        Layers.isChildLayer(Layers.access(getCurrentPage(state), indexPath)),
+      )
+      // Reverse the indexPaths to simplify deletion
+      .reverse()
+  );
+};
+
+const createGroup = (
+  action: 'group' | 'symbolMaster',
```

Comment:
Let's pass in the `Models.symbolMaster` and `Models.group` instead of this action-based approach. That way we can use a generic type:

```ts
const createGroup = <T extends Sketch.Group | Sketch.SymbolMaster>(
  layer: T,
  page: Sketch.Page,
  ids: string[],
  name: string,
  indexPaths: IndexPath[],
): T | undefined => { ... }
```

We can move the `symbolID` assignment outside of this function

Reply:
sounds good!